controllers/openshift_controller.go

```diff
@@ -595,212 +390,162 @@ func (r *KataConfigOpenShiftReconciler) processKataConfigDeleteRequest() (ctrl.R
-			if err != nil {
-				return ctrl.Result{}, err
-			}
+	r.Log.Info("Making sure parent MCP is synced properly, KataNodeRole=" + machinePool)
+	r.kataConfig.Status.UnInstallationStatus.InProgress.IsInProgress = corev1.ConditionTrue
+	if _, ok := r.kataConfig.Spec.KataConfigPoolSelector.MatchLabels["node-role.kubernetes.io/"+machinePool]; ok {
+		mc, err := r.newMCForCR(machinePool)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		var isMcDeleted bool
+
+		err = r.Client.Get(context.TODO(), types.NamespacedName{Name: mc.Name}, mc)
+		if err != nil && k8serrors.IsNotFound(err) {
+			isMcDeleted = true
 		} else if err != nil {
 			return ctrl.Result{}, err
 		}
 
-		if r.kataConfig.Status.UnInstallationStatus.Completed.CompletedNodesCount != r.kataConfig.Status.TotalNodesCount {
-			r.Log.Info("KataConfig uninstallation: ", "Number of nodes completed uninstallation ",
-				r.kataConfig.Status.UnInstallationStatus.Completed.CompletedNodesCount,
-				"Total number of kata installed nodes ", r.kataConfig.Status.TotalNodesCount)
-			// TODO - we don't need this nil check if we know that pool is always initialized
-			if r.kataConfig.Spec.KataConfigPoolSelector != nil &&
-				r.kataConfig.Spec.KataConfigPoolSelector.MatchLabels != nil && len(r.kataConfig.Spec.KataConfigPoolSelector.MatchLabels) > 0 {
-				if r.clientset == nil {
-					r.clientset, err = getClientSet()
-					if err != nil {
-						return ctrl.Result{}, err
-					}
-				}
-
-				for _, nodeName := range r.kataConfig.Status.UnInstallationStatus.InProgress.BinariesUnInstalledNodesList {
-					if contains(r.kataConfig.Status.UnInstallationStatus.Completed.CompletedNodesList, nodeName) {
-						continue
-					}
-
-					if _, ok := r.kataConfig.Spec.KataConfigPoolSelector.MatchLabels["node-role.kubernetes.io/"+machinePool]; !ok {
-						r.Log.Info("Removing the kata pool selector label from the node", "node name ", nodeName)
-						node, err := r.clientset.CoreV1().Nodes().Get(context.Background(), nodeName, metav1.GetOptions{})
-						if err != nil {
-							return ctrl.Result{}, err
-						}
-
-						nodeLabels := node.GetLabels()
-
-						for k := range r.kataConfig.Spec.KataConfigPoolSelector.MatchLabels {
-							delete(nodeLabels, k)
-						}
-
-						node.SetLabels(nodeLabels)
-						_, err = r.clientset.CoreV1().Nodes().Update(context.Background(), node, metav1.UpdateOptions{})
-
-						if err != nil {
-							return ctrl.Result{}, err
-						}
-					}
-				}
+		if !isMcDeleted {
+			err = r.Client.Delete(context.TODO(), mc)
+			if err != nil {
+				// error during removing mc, don't block the uninstall. Just log the error and move on.
+				r.Log.Info("Error found deleting machine config. If the machine config exists after installation it can be safely deleted manually.",
+					"mc", mc.Name, "error", err)
 			}
+			// Sleep for MCP to reflect the changes
+			r.Log.Info("Pausing for a minute to make sure worker mcp has started syncing up")
+			time.Sleep(60 * time.Second)
```

Comment:
It's a good pint and yes, I think it could be replaced with e.g. a Watch on the machine config pool. But since this has been used since a looong time and is not new in this PR I would prefer to do it in another PR. Do you agree?

Reply:
I've opened https://github.com/openshift/sandboxed-containers-operator/issues/87 to track such cases, and I'm totally fine on having this as a different effort / different PR.

Reply:
From my - admittedly very limited - experience I think this is a common pattern in controller programming - if something turns out not to be ready just try again in x seconds.  Not as elegant or efficient as an async solution but dead simple, and since the wait periods tend to be in (tens of) seconds it might not be worth the added complexity to sweat the efficiency...